libs/utils/src/postgres_backend.rs

```diff
@@ -93,48 +183,54 @@ impl fmt::Display for AuthType {
     }
 }
 
-#[derive(Clone, Copy)]
-pub enum ProcessMsgResult {
-    Continue,
-    Break,
+/// Either full duplex Framed or write only half; the latter is left in
+/// PostgresBackend after call to `split`. In principle we could always store a
+/// pair of splitted handles, but that would force to to pay splitting price
+/// (Arc and kinda mutex inside polling) for all uses (e.g. pageserver).
+enum MaybeWriteOnly {
+    Full(Framed<MaybeTlsStream>),
+    WriteOnly(FramedWriter<WriteHalf<MaybeTlsStream>>),
+    Broken, // temporary value palmed off during the split
 }
 
-/// Always-writeable sock_split stream.
-/// May not be readable. See [`PostgresBackend::take_stream_in`]
-pub enum Stream {
-    Bidirectional(BidiStream),
-    WriteOnly(WriteStream),
-}
+impl MaybeWriteOnly {
+    async fn read_message(&mut self) -> Result<Option<FeMessage>, ConnectionError> {
+        match self {
+            MaybeWriteOnly::Full(framed) => framed.read_message().await,
+            MaybeWriteOnly::WriteOnly(_) => {
+                Err(io::Error::new(ErrorKind::Other, "reading from write only half").into())
+            }
+            MaybeWriteOnly::Broken => panic!("IO on invalid MaybeWriteOnly"),
```

Comment:
I don't care much and I'll do that if you insist, however I actually find the current state slightly better in terms of readability. IO in `Broken` can happen only if there is an blunder in relatively compact piece of *internal* code inside postgres_backend. It just should never happen. `Broken` wouldn't even exist has rust been more intelligent in handing these replacement ; it should allow to temporary move out struct field if no error can happen before it is moved back. However, attempt to read from write only half is quite possible in case of external user error, and you might even do that sorta intentional in dirty code (like we had for a long time before this patch).

Reply:
citing [too many linked lists](https://rust-unofficial.github.io/too-many-lists/first-push.html) "In principle, this is something Rust could actually accept, but it won't (for various reasons -- the most serious being [exception safety](https://doc.rust-lang.org/nightly/nomicon/exception-safety.html))"

Reply:
I do not insist on removing broken variant. I only insist on choosing one strategy to handle invariant violation. I'm fine with either Result or panic. I dont see why we should use both to indicate the same thing.

I can see one possible direction on how Broken can be avoided, but as we agreed, it is not an easy way.

Reply:
I explained above why I intentionally chose to have two different strategies here, and why `WriteOnly` and `Broken` are different things. We can have a call if that wasn't clear.

Reply:
Hm, or you are talking about bail! in `Broken` state at `start_tls`? I'll replace it with panic.

Reply:
> Hm, or you are talking about bail! in Broken state at start_tls?

Exactly.

Reply:
fixed